File: week9/p2.c
#include <stdio.h>
#include <stdlib.h>

struct Node {
    int key;
    struct Node* next;
};

struct Hash {
    int bucket_count;
    struct Node** table;
};

struct Hash* createHash(int buckets) {
    struct Hash* h = (struct Hash*)malloc(sizeof(struct Hash));
    h->bucket_count = buckets;
    h->table = (struct Node**)malloc(buckets * sizeof(struct Node*));
    for (int i = 0; i < buckets; i++) {
        h->table[i] = NULL;
    }
    return h;
}

int getHashIndex(struct Hash* h, int key) {
    return key % h->bucket_count;
}

void insert(struct Hash* h, int key) {
    int index = getHashIndex(h, key);
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    newNode->key = key;
    newNode->next = h->table[index];
    h->table[index] = newNode;
}

void removeKey(struct Hash* h, int key) {
    int index = getHashIndex(h, key);
    struct Node* curr = h->table[index];
    struct Node* prev = NULL;
    while (curr != NULL) {
        if (curr->key == key) {
            if (prev == NULL) {
                h->table[index] = curr->next;
            } else {
                prev->next = curr->next;
            }
            free(curr);
            return;
        }
        prev = curr;
        curr = curr->next;
    }
}

void display(struct Hash* h) {
    for (int i = 0; i < h->bucket_count; i++) {
        printf("%d", i);
        struct Node* curr = h->table[i];
        while (curr != NULL) {
            printf(" --> %d", curr->key);
            curr = curr->next;
        }
        printf("\n");
    }
}

int main() {
    int keys[] = {7, 18, 12, 25};
    int n = sizeof(keys) / sizeof(keys[0]);

    struct Hash* hash_table = createHash(7);

    for (int i = 0; i < n; i++) {
        insert(hash_table, keys[i]);
    }

    removeKey(hash_table, 12);
    display(hash_table);

    return 0;
}

File: week9/p2.py
class Hash:
    def __init__(self, buckets):
        self.bucket_count = buckets
        self.table = [[] for _ in range(self.bucket_count)]

    def insert(self, key):
        index = self.get_hash_index(key)
        self.table[index].append(key)

    
    def remove(self, key):
        index = self.get_hash_index(key)
        if key in self.table[index]:
            self.table[index].remove(key)

    def display(self):
        for i in range(self.bucket_count):
            print(i, end="")
            for key in self.table[i]:
                print(" -->", key, end="")
            print()

    def get_hash_index(self, key):
        return key % self.bucket_count

if __name__ == "__main__":
    keys = [7, 18, 12, 25]

    hash_table = Hash(7)

    for key in keys:
        hash_table.insert(key)

    hash_table.remove(12)
    hash_table.display()